work_queue/src/work_queue.c

```diff
@@ -851,31 +847,76 @@ static int get_transfer_wait_time(struct work_queue *q, struct work_queue_worker
 	return timeout;
 }
 
-static void remove_factory(struct work_queue_factory_info *f)
+static int factory_trim_workers(struct work_queue *q, struct work_queue_factory_info *f)
+{
+	// Iterate through all workers and shut idle ones down
+	char *factory_name = f->name;
+	struct work_queue_worker *w;
+	char *key;
+	int i = 0;
+
+	hash_table_firstkey(q->worker_table);
+	while( f->connected_workers > f->max_workers &&
+			hash_table_nextkey(q->worker_table, &key, (void **) &w) ) {
+		if( !strcmp(factory_name, w->factory_name) &&
+			!itable_size(w->current_tasks) ) {
+			shut_down_worker(q, w);
+			i++;
+		}
+	}
+
+	// Trimmed some but not enough workers.
+	if (f->connected_workers > f->max_workers) f->trimmed = 1;
```

Comment:
f->trimmed should be set to 1, regardless.

Reply:
The notion here is:

- After trimming the workers, if the we succeed to reduce the number of workers as desired, we consider this round of draining is completed. So we should actually restore `trimmed` to 0.
- If we didn't succeed to reduce that many workers, we set `trimmed = 1`, so that in this round of draining, we won't attempt to trim it again.

In short, `trimmed` is just for a single round of draining. If the draining is completed, we should set it back to 0.

Reply:
Consider the following case:

`connected_workers = 15`, `6` of them are idle, `max_workers = 10`. This will trigger draining, and after trimming, `10` workers are left, `1` of them is still idle.

Then, one more worker connected and starts to work. `max_workers` is not changed, but this will trigger draining again.

If we use `if (f->connected_workers > f->max_workers) f->trimmed = 1;`, `trimmed` will not be set in first round of trimming, and we are able to trim one more idle worker in the second round.

If we use `f->trimmed = 1;`, `trimmed` will be set in the first round, thus we won't be able to trim that one last idle worker in the second round.

Reply:
Ah good point. I would like to eliminate any such hidden interactions by reducing the state we keep, so I suggest:
- Remove f->trimmed completely.
- Trigger a factory trim only when max_workers is reduced for that factory.
- When a new worker sends the "info factory..." message, immediately shutdown the worker if `connected_workers >= max_workers`
- Do not schedule tasks to workers of factories with `connected_workers > max_workers`
- On task retrieval, if no other current task in the worker, shutdown the worker if `connected_workers > max_workers`

Reply:
Sounds good, the first three points need to be done, last two are already written.